lua/rip-substitute/rg-operations.lua

```diff
@@ -91,113 +115,131 @@ function M.executeSubstitution()
-	state.matchCount = 0
-	vim.api.nvim_buf_clear_namespace(state.targetBuf, state.incPreviewNs, 0, -1)
-	local hlGroup = require("rip-substitute.config").config.incrementalPreview.matchHlGroup
-
-	local toSearch, toReplace = M.getSearchAndReplaceValuesFromPopup()
-	if toSearch == "" then return end
-
-	-- DETERMINE MATCHES
-	local rgArgs = { toSearch, "--line-number", "--column", "--only-matching" }
-	local code, searchMatches = runRipgrep(rgArgs)
-	if code ~= 0 then return end
-
-	-- RANGE: FILTER MATCHES
-	-- PERF For single files, `rg` gives us results sorted by lines, so we can
-	-- `slice` instead of `filter` to improve performance.
+--- RANGE: FILTER MATCHES
+--- PERF For single files, `rg` gives us results sorted by lines, so we can
+--- `slice` instead of `filter` to improve performance.
+---@param matches RipSubstituteMatch
+---@param range CmdRange
+local function getMatchesInRange(matches, range)
 	local rangeStartIdx, rangeEndIdx
-	if state.range then
-		for i = 1, #searchMatches do
-			local lnum = tonumber(searchMatches[i]:match("^(%d+):"))
-			local inRange = lnum >= state.range.start and lnum <= state.range.end_
-			if rangeStartIdx == nil and inRange then rangeStartIdx = i end
-			if rangeStartIdx and lnum > state.range.end_ then
-				rangeEndIdx = i - 1
-				break
-			end
+	for i, match in ipairs(matches) do
+		local inRange = match.row >= range.start and match.row <= range.end_
+		if rangeStartIdx == nil and inRange then rangeStartIdx = i end
+		if rangeStartIdx and match.row > range.end_ then
+			rangeEndIdx = i - 1
+			break
 		end
-		if rangeStartIdx == nil then return end -- no matches in range
-		searchMatches = vim.list_slice(searchMatches, rangeStartIdx, rangeEndIdx)
 	end
+	if rangeStartIdx == nil then return end -- no matches in range
+	return vim.list_slice(matches, rangeStartIdx, rangeEndIdx)
+end
 
-	state.matchCount = #searchMatches
-
+---@param matches RipSubstituteMatch
+---@param viewStartLnum number
+---@param viewEndLnum number
+---@return number|nil, number | nil
+local function getViewportRange(matches, viewStartLnum, viewEndLnum)
 	-- VIEWPORT: FILTER MATCHES
 	local viewStartIdx, viewEndIdx
-	for i = 1, #searchMatches do
-		local lnum = tonumber(searchMatches[i]:match("^(%d+):"))
-		if not viewStartIdx and lnum >= viewStartLnum and lnum <= viewEndLnum then
+	for i, match in ipairs(matches) do
+		if not viewStartIdx and match.row >= viewStartLnum and match.row <= viewEndLnum then
 			viewStartIdx = i
 		end
-		if viewStartIdx and lnum > viewEndLnum then
+		if viewStartIdx and match.row > viewEndLnum then
 			viewEndIdx = i - 1
 			break
 		end
 	end
-	if not viewStartIdx then return end -- no matches in viewport
-	if not viewEndIdx then viewEndIdx = #searchMatches end -- viewport is at end of file
+end
+
+---@param match RipSubstituteMatch
+---@param selected boolean
+---@param targetBuf number
+---@param incPreviewNs number
+function M.highlightReplacement(match, selected, targetBuf, incPreviewNs)
```

Comment:
`targetBuf` and `incPreviewNs` do not need to be passed, since they can be accessed via `state`.